discovery_engine/lib/src/infrastructure/repository/hive_source_preference_repo.dart

```diff
@@ -0,0 +1,43 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import 'package:hive/hive.dart' show Hive, Box;
+import 'package:xayn_discovery_engine/discovery_engine.dart';
+import 'package:xayn_discovery_engine/src/domain/models/source_preference.dart';
+import 'package:xayn_discovery_engine/src/domain/repository/source_preference_repo.dart';
+import 'package:xayn_discovery_engine/src/infrastructure/box_name.dart'
+    show sourcePreferenceBox;
+
+class HiveSourcePreferenceRepository implements SourcePreferenceRepository {
+  Box<SourcePreference> get box => Hive.box<SourcePreference>(sourcePreferenceBox);
+
+  @override
+  Iterable<Source> getFavorites() {
+    return box.values
+        .where((el) => el.mode == PreferenceMode.favorite)
+        .map((el) => el.source);
+  }
+
+  @override
+  Iterable<Source> getExcluded() {
+    return box.values
+        .where((el) => el.mode == PreferenceMode.excluded)
+        .map((el) => el.source);
+  }
```

Comment:
But the Iterable would retain a reference to the function inside map, right? In what scenario would it not be executed? Is this something about Dart's runtime model that I can read up on?

Reply:
> But the Iterable would retain a reference to the function inside map, right?
yes

> In what scenario would it not be executed?
if the returned elements are not iterated. So they need to be for example converted with `.toList()` or `.toSet()`

Most of the time it's nothig to worry about, but sometimes it can also lead to some performance issues. I've send you a message on priv about that :)